### PR TITLE
Improve method completion to work `prevind("θ",<tab>.

### PR DESCRIPTION
**Pull request by dhoegh**
I have improved method completion to also work when `max(1,<tab>` and it will only show the methods that matches the type of the given args. If the type of the arguments can’t be determined it does not return any suggestions. Example of use:

```
a = [1, 1]
julia> max(a, <tab>
max{T1<:Real,T2<:Real}(::AbstractArray{T1<:Real,N},::T2<:Real) at operators.jl:3
72
max{T1<:Real,T2<:Real}(::AbstractArray{T1<:Real,N},::AbstractArray{T2<:Real,N})
at operators.jl:376
max(x,y) at operators.jl:56
max(a,b,c) at operators.jl:83
max(a,b,c,xs...) at operators.jl:84
julia> max(a, 
```

But it do not return anything by doing:

```
julia>max([1, 2],<tab>
```

For further examples, see the test cases.

### CODE REVIEW

**Comment by dhoegh** [comment]
An alternative to when the type of the arguments can’t be determined, it could sort the matches in the same manner as in: #9485. Where it counts the number of matching arguments and then displays the top 3. best matching methods.

**Comment by dhoegh** [comment]
Bump any interest in this?

**Comment by timholy** [comment]
:+1:

**Comment by StefanKarpinski** [comment]
Yes, it's very nice. @timholy, if you've tested, approve and want to merge, go for it.

**Comment by ivarne** [comment]
+1

The idea looks really nice, but I haven't tried it to see if it works correctly.

**Comment by timholy** [comment]
Likewise, I'm just voicing my approval for the idea. Looks like @Keno has self-assigned, and he knows rather more about REPL code than I do.

**Comment by dhoegh** [comment]
I have made some small corrections, since there is an interest in the pull-request.

**Comment by dhoegh** [comment]
Bump.

**Comment by timholy** [comment]
@Keno or @blakejohnson, could one of you review this? @dhoegh, if you don't get a response in a couple days, bump again and I'll merge it.

File: base/REPLCompletions.jl [hunk 1]
```diff
@@ -134,16 +134,68 @@ function complete_path(path::AbstractString, pos)
...
 
-function complete_methods(input::AbstractString)
-    tokens = split(input, '.')
-    fn = Main
-    for token in tokens
-        sym = symbol(token)
-        isdefined(fn, sym) || return UTF8String[]
-        fn = fn.(sym)
+# Determines whether a string endswiths ',' or '(' when disregarding whitespace_chars
+function ends_c(s::AbstractString)
```
Comment (blakejohnson) [inline base/REPLCompletions.jl]: This function name is pretty opaque to me. Can we come up with something more descriptive?

File: base/REPLCompletions.jl [hunk 2]
```diff
@@ -134,16 +134,68 @@ function complete_path(path::AbstractString, pos)
...
+        end
+        braces == 1 && break
+    end
+    braces != 1 && return 0:-1
+    startind = nextind(s, rsearch(s, non_identifier_chars, reverseind(r, i)))
+    return startind:endof(s)
+end
+
+# Return the next value when the sym is defined in current namespace fn
+function return_val(sym::Expr, fn)
```
Comment (blakejohnson) [inline base/REPLCompletions.jl]: Both the comment and the function name are again unhelpful in describing what this does.

**Comment by blakejohnson** [comment]
The behavior is nice and I see no problems in terms of interaction with the REPL. I always find expression parsing code to be a bit impenetrable, but I think with a couple name changes to help future readers of this code, it is good to go.

**Comment by dhoegh** [comment]
@blakejohnson I have made some method renaming and commenting. I have also fixed the same bug as I discovered in #10129.

**Comment by dhoegh** [comment]
@blakejohnson thank you for the review, I do agree that expression parsing is difficult to comprehend.

**Comment by blakejohnson** [comment]
Thanks, that is a big improvement in readability for me.

**Comment by blakejohnson** [comment]
@dhoegh looks like there are some new test failures.

**Comment by dhoegh** [comment]
Yes, I changed how the range is returned and did not get the test case corrected. I will take a look at it tomorrow morning.

**Comment by dhoegh** [comment]
I have updated the tests and added some extra, the test passes Travis and appveyor 64 bit and the 32-bit failure on appveyor is unrelated.

**Comment by blakejohnson** [comment]
Great, thank you @dhoegh.

**Comment by timholy** [comment]
Thanks to both of you!

**Pull request merged**